zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -1,49 +1,161 @@
-use crate::evm_circuit::{
-    param::MAX_MEMORY_SIZE_IN_BYTES,
-    util::{
-        constraint_builder::ConstraintBuilder,
-        math_gadget::{ConstantDivisionGadget, MaxGadget},
-        Address, MemorySize,
+use crate::{
+    evm_circuit::{
+        param::{
+            N_BYTES_GAS, N_BYTES_MEMORY_ADDRESS, N_BYTES_MEMORY_WORD_SIZE,
+        },
+        util::{
+            constraint_builder::ConstraintBuilder,
+            from_bytes,
+            math_gadget::IsZeroGadget,
+            math_gadget::{ConstantDivisionGadget, MinMaxGadget},
+            sum, Cell, MemoryAddress, Word,
+        },
     },
+    util::Expr,
+};
+use array_init::array_init;
+use bus_mapping::{
+    eth_types::{ToLittleEndian, U256},
+    evm::GasCost,
 };
-use crate::util::Expr;
-use bus_mapping::evm::GasCost;
 use halo2::plonk::Error;
 use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Expression};
+use std::convert::TryInto;
 
 /// Decodes the usable part of an address stored in a Word
 pub(crate) mod address_low {
     use crate::evm_circuit::{
-        param::NUM_ADDRESS_BYTES_USED,
-        util::{from_bytes, Address, Word},
+        param::N_BYTES_MEMORY_ADDRESS,
+        util::{from_bytes, Word},
     };
     use halo2::{arithmetic::FieldExt, plonk::Expression};
 
     pub(crate) fn expr<F: FieldExt>(address: &Word<F>) -> Expression<F> {
-        from_bytes::expr(&address.cells[0..NUM_ADDRESS_BYTES_USED])
+        from_bytes::expr(&address.cells[..N_BYTES_MEMORY_ADDRESS])
     }
 
-    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> Address {
-        from_bytes::value::<F>(&address[0..NUM_ADDRESS_BYTES_USED])
-            .get_lower_128() as Address
+    pub(crate) fn value(address: [u8; 32]) -> u64 {
+        let mut bytes = [0; 8];
+        bytes.copy_from_slice(&address[..N_BYTES_MEMORY_ADDRESS]);
+        u64::from_le_bytes(bytes)
     }
 }
 
 /// The sum of bytes of the address that are unused for most calculations on the
 /// address
 pub(crate) mod address_high {
     use crate::evm_circuit::{
-        param::NUM_ADDRESS_BYTES_USED,
+        param::N_BYTES_MEMORY_ADDRESS,
         util::{sum, Word},
     };
     use halo2::{arithmetic::FieldExt, plonk::Expression};
 
     pub(crate) fn expr<F: FieldExt>(address: &Word<F>) -> Expression<F> {
-        sum::expr(&address.cells[NUM_ADDRESS_BYTES_USED..32])
+        sum::expr(&address.cells[N_BYTES_MEMORY_ADDRESS..])
     }
 
     pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> F {
-        sum::value::<F>(&address[NUM_ADDRESS_BYTES_USED..32])
+        sum::value::<F>(&address[N_BYTES_MEMORY_ADDRESS..])
+    }
+}
+
+/// Convert the dynamic memory offset and length from random linear combiation
+/// to integer. It handles the "no expansion" feature when length is zero.
+#[derive(Clone, Debug)]
+pub(crate) struct MemoryAddressGadget<F> {
+    memory_offset: Cell<F>,
```

Comment:
`memory_offset` is starting point of address for expansion ?

Reply:
Yes, more precisely, the starting point of address for `call_data` or `return_data` specified by caller.

Reply:
My understanding of this gadget is that it aims to provide the max used address after a memory access by an opcode that takes `offset` and `length` like `CALLDATACOPY`, `CODECOPY`, `EXTCODECOPY`, `RETURNDATACOPY`.  And I think this max address would then be used in the `MemoryExpansionGadget` to calculate the `gas_cost` of the possible expansion.

And `MemoryAddressGadget` returns `max_address = 0` when the `length = 0` so that no memory expansion is performed in such case, even if the offset is past the current max used address.

Reply:
Exactly, thanks for explanation!